reference-implementation/lib/writable-stream.js

```diff
@@ -4,225 +4,661 @@ const { InvokeOrNoop, PromiseInvokeOrNoop, PromiseInvokeOrFallbackOrNoop, Valida
-      this._started = true;
-      this._startedPromise = undefined;
-    });
-    this._startedPromise.catch(r => ErrorWritableStream(this, r)).catch(rethrowAssertionErrorRejection);
+    this._writableStreamController = new WritableStreamDefaultController(this, underlyingSink, size, highWaterMark);
   }
 
-  get closed() {
-    if (!IsWritableStream(this)) {
-      return Promise.reject(new TypeError('WritableStream.prototype.closed can only be used on a WritableStream'));
+  get locked() {
+    if (IsWritableStream(this) === false) {
+      throw CreateWritableStreamBrandCheckException('locked');
     }
 
-    return this._closedPromise;
+    return IsWritableStreamLocked(this);
   }
 
-  get state() {
-    if (!IsWritableStream(this)) {
-      throw new TypeError('WritableStream.prototype.state can only be used on a WritableStream');
+  abort(reason) {
+    if (IsWritableStream(this) === false) {
+      return Promise.reject(CreateWritableStreamBrandCheckException('abort'));
+    }
+
+    if (IsWritableStreamLocked(this) === true) {
+      return Promise.reject(new TypeError('Cannot abort a stream that already has a reader'));
     }
 
-    return this._state;
+    WritableStreamAbort(this, reason);
   }
 
-  abort(reason) {
-    if (!IsWritableStream(this)) {
-      return Promise.reject(new TypeError('WritableStream.prototype.abort can only be used on a WritableStream'));
+  getWriter() {
+    if (IsWritableStream(this) === false) {
+      throw CreateWritableStreamBrandCheckException('getWriter');
     }
 
-    if (this._state === 'closed') {
-      return Promise.resolve(undefined);
-    }
-    if (this._state === 'errored') {
-      return Promise.reject(this._storedError);
+    return AcquireWritableStreamDefaultWriter(this);
+  }
+}
+
+exports.WritableStream = WritableStream;
+
+// Helper functions for the WritableStream.
+
+function CreateWritableStreamBrandCheckException(name) {
+  return new TypeError('WritableStream.prototype.' + name + ' can only be used on a WritableStream')
+}
+
+// Abstract operations for the WritableStream.
+
+function AcquireWritableStreamDefaultWriter(stream) {
+  return new WritableStreamDefaultWriter(stream);
+}
+
+function IsWritableStream(x) {
+  if (!typeIsObject(x)) {
+    return false;
+  }
+
+  if (!Object.prototype.hasOwnProperty.call(x, '_writableStreamController')) {
+    return false;
+  }
+
+  return true;
+}
+
+function IsWritableStreamLocked(stream) {
+  assert(IsWritableStream(stream) === true, 'IsWritableStreamLocked should only be used on known writable streams');
+
+  if (stream._writer === undefined) {
+    return false;
+  }
+
+  return true;
+}
+
+function WritableStreamAbort(stream, reason) {
+  const state = stream._state;
+  if (state === 'closed') {
+    return Promise.resolve(undefined);
+  }
+  if (state === 'errored') {
+    return Promise.reject(stream._storedError);
+  }
+
+  assert(state === 'writable' || state === 'waiting' || state === 'closing');
```

Comment:
Basically my concern is with keeping multiple pieces of information coordinated. For example maybe what we want here is to check if the ready promise exists (or to make WritableStreamDefaultWriterReadyPromiseResolve do nothing if it doesn't exist).